### PR TITLE
Dbex 64564 lighthouse supporting evidence attachment

### PR DESCRIPTION
**Pull request by NB28VT**
## Summary
* Part of the wider EVSS to Lighthouse migration
* Updates all document-related behavior around Form 526 submissions to use the new [Lighthouse Benefits Documents API](https://dev-developer.va.gov/explore/api/benefits-documents/docs?version=current)
* Most of these changes are gated behind the `disability_compensation_lighthouse_document_service_provider` feature flag

### Current Changes Overview

1. Creates a new `LighthouseSupportingEvidenceAttachment` class that inherits from `FormAttachment`. This was required to solve an inheritance issue that is well documented in the [commit message](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/7e5c4658bb3877539f4f9fcad82cae95f5421bcd) for this change. Note I also had to create Lighthouse verisons of the superclasses and associated services for this attachment subtype, which mirrors the behavior in `SupportingEvidenceAttachment`

2. Implements a new [`BenefitsDocuments::Form526LighthouseDocumentsService`](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/99a2865c42175c136e0b5c45ca199ae2ae9cd0e1) for interacting with the `BenefitsDocuments::WorkerService` implemented by the mobile team. Note that while is very similar to the [service implemented by the mobile team for this purpose](https://github.com/department-of-veterans-affairs/vets-api/blob/dbex-64564-lighthouse-supporting-evidence-attachment/app/sidekiq/lighthouse/document_upload.rb), our team is in the process of improving our logging and tracking and we preferred to log this behavior using the [`wrap_with_logging` helper](https://github.com/department-of-veterans-affairs/vets-api/blob/dbex-64564-lighthouse-supporting-evidence-attachment/lib/logging/third_party_transaction.rb). Therefore it was easiest to create our own separate service.

3. For backwards compatibility, [updates the create route](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/a47dddf041b4a1f73537f2d70be28929b9245952) veterans submit their own documents to in the `V0::UploadSupportingEvidencesController` to save a `LighthouseSupportingEvidenceAttachment` in addition to the `SupportingEvidenceAttachment` that gets created now. Note this is not feature gated since we want to start creating these now. More information on the reasoning for this change is in the commit message.

4. Updates the `EVSS::DisabilityCompensationForm::SubmitUploads` worker to use the Lighthouse Documents API. This is the job that submits additional documents the veteran provided when they were filling out Form 526 (medical forms, assault investigation documentation etc.). It runs once per document the veteran uploaded.

5. Updates the `EVSS::DisabilityCompensationForm::SubmitForm0781` worker to use the Lighthouse Documents API. This is the job that submits a PDF of Form 0781 (PTSD claim) and Form 0781a (sexual assault) if they were included in the application.

6. Updates the `EVSS::DisabilityCompensationForm::UploadBddInstructions` worker to use the Lighthouse Documents API. This is the job that submits a static PDF we always send with Form 526 submissions called "BDD Instructions".

#### Additional Notes

I set up my own Lighthouse Sandbox credentials to allow me to run the worker tests with the live endpoint, in order to record VCR cassette fixtures of the response. It appears in all cases this worked succesfully.

Note I did not make any changes to the [`EVSS::DisabilityCompensationForm:SubmitForm8940` worker](https://github.com/department-of-veterans-affairs/vets-api/blob/dbex-64564-lighthouse-supporting-evidence-attachment/lib/evss/disability_compensation_form/form8940.rb) to add Lighthouse support. My understanding is 8940 is a form for documenting prior employment as part of documenting unemployability. It's in the codebase and accessible within the Form 526 flow in staging, but it's not in production. This is because it's a feature that was never fully implemented. So it was cut out of the scope for this migration.

## Testing done
* TBD: the next stage of this project will be writing a comprehensive rollout plan that will include any testing. Note I have been successfully working with the Lighthouse Sandbox endpoint and have access to the VBMS claims portal in the testing environment, so some manual testing should be possible before we incrementally release this using the Flipper switch.

## What areas of the site does it impact?
* All changes are scoped to the Form 526 claim workflow. 

## Acceptance criteria

- [X]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [X]  No error nor warning in the console.
- [X]  Events are being sent to the appropriate logging solution
- [X]  Documentation has been updated (link to documentation) N/A
- [X]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [X]  Feature/bug has a monitor built into Datadog or Grafana (if applicable)
- [X]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [X]  I added a screenshot of the developed feature N/A

## Requested Feedback

Curious about my approach to updating the workers. I tried to make these changes in a way that:

1. Has minimal impact on the existing code to prevent breaking anything
2. Makes it easy to remove the EVSS behavior once we release this feature for all users.
3. Allows us to use the `wrap_with_logging` helper as close to the interaction with the Lighthouse API as possible.

### CODE REVIEW

**Comment by NB28VT** [comment]
@austinjprice one open question I have you may know how to get an answer to:

There are different documents that get uploaded during the Form 526 submission process, do those all need to be migrated over to Lighthouse in the same launch or if some of them still go to EVSS will everything still end up in the right place?

**Comment by austinjprice** [comment]
> @austinjprice one open question I have you may know how to get an answer to:
> 
> There are different documents that get uploaded during the Form 526 submission process, do those all need to be migrated over to Lighthouse in the same launch or if some of them still go to EVSS will everything still end up in the right place?

@NB28VT I can run with it! Two questions: 1. can you give me examples of different documents? 2. Is the question to allow for incremental release with lower risk?

File: .github/CODEOWNERS
Comment (freeheeling) [inline .github/CODEOWNERS:1046]: @NB28VT – Did you intentionally remove this file and owner group? There's a failing check related to this change because now there are 2 owners without a corresponding file.

File: .github/CODEOWNERS
Comment (freeheeling) [inline .github/CODEOWNERS:1937]: @NB28VT – I'm not certain on the protocol for updates to this file, but it appears, for the most part, that additions are listed alphabetically by file path.

File: app/controllers/v0/upload_supporting_evidences_controller.rb
Comment (freeheeling) [inline app/controllers/v0/upload_supporting_evidences_controller.rb:9]: Is this constant being accessed outside this controller?

File: app/controllers/v0/upload_supporting_evidences_controller.rb
Comment (freeheeling) [inline app/controllers/v0/upload_supporting_evidences_controller.rb:7]: Would it not be possible to continue utilizing this concern and conditionally call `save_lighthouse_attachment` on `create` if `FORM_ATTACHMENT_MODEL == SupportingEvidenceAttachment`?

File: spec/controllers/v0/upload_supporting_evidences_controller_spec.rb
Comment (freeheeling) [inline spec/controllers/v0/upload_supporting_evidences_controller_spec.rb:1]: Deleted intentionally? Controller is still present.

File: .github/CODEOWNERS
Comment (SamStuckey) [reply .github/CODEOWNERS:1046]: + 1, this is basically a syntax error @NB28VT , the `@` symbol here must be proceeded by a filepath

File: app/controllers/v0/upload_supporting_evidences_controller.rb
Comment (SamStuckey) [reply app/controllers/v0/upload_supporting_evidences_controller.rb:7]: the problem is `FormAttachmentCreate` inherits has some complex inheritance that defines what service it faces (EVSS or lighthouse) at the time of class definition, aka application load.  Therefor we had to remove some nominally harmless dependancies that were actually obfuscating downstream coupling ( I think)

File: app/models/lighthouse_supporting_evidence_attachment.rb
Comment (SamStuckey) [inline app/models/lighthouse_supporting_evidence_attachment.rb:5]: Is this a replacement for an EVSS object?

File: app/sidekiq/evss/disability_compensation_form/submit_form0781.rb
Comment (SamStuckey) [inline app/sidekiq/evss/disability_compensation_form/submit_form0781.rb:17]: 👌

File: app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb
Comment (SamStuckey) [inline app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb:6]: this could all be replaced with the logging wrapper, fyi.

```
wrap_with_logging(
  :store,
      additional_instance_logs: {
        upload_file: %i[size headers]
      }
    )
```
should do the same thing

File: app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb
Comment (SamStuckey) [inline app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb:33]: if you use my logging wrapper you can remove these methods

File: lib/lighthouse/benefits_documents/form_526_lighthouse_documents_service.rb
Comment (NB28VT) [inline lib/lighthouse/benefits_documents/form_526_lighthouse_documents_service.rb:10]: @SamStuckey quick questions here as I get into the DataDog dashboards.

 I'm thinking I should be monitoring the response code and body returned from this `upload_from_client` method, along with the `document_type` argument passed to `upload_lighthouse_document`. Assume there is a way to create a widget for each of the document types I'm using this method for:
- Supplemental documents (the ones the vet uploads that are handled in the SubmitUploads worker)
- Form 0781
- Form 0781a
- BDD Instructions

So initial questions:
1. Do you have an example of a dashboard you've created that leverages your 3PI logging behavior?
2. For setting up dashboards for all of those use cases (recall this is a module I'm reusing in a few places), would it make more sense to update the code here to this:

```ruby
  wrap_with_logging(
      :upload_from_client
      additional_instance_logs: {
        document_type: [:document_type]
      }
    )
```

where `document_type` becomes a new instance var, and set up the widget that way.

OR

Is there a way in DataDog to set up a widget that tracks this by caller? Eg. `EVSS::DisabilityCompensationForm::SubmitForm0781#upload_lighthouse_document` would be one.

LMK if this would be easier to chat about in a call, but basically wondering if I should make changes here for the purposes of my monitoring. I'm guessing the second option is easier and makes more sense.

Comment (NB28VT) [reply lib/lighthouse/benefits_documents/form_526_lighthouse_documents_service.rb:10]: (also for reference if you follow the below code all the way through the Lighthouse API wrapper I'm using the mobile team built, this returns a `Faraday::Response` object)
```
    def upload_from_client(user_icn, file_body, lighthouse_document)
      client = BenefitsDocuments::WorkerService.new(user_icn)
      client.upload_document(file_body, lighthouse_document)
    end
```

File: app/controllers/v0/upload_supporting_evidences_controller.rb
Comment (NB28VT) [reply app/controllers/v0/upload_supporting_evidences_controller.rb:7]: Yup, Sam's recollection of this is correct, see https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/7e5c4658bb3877539f4f9fcad82cae95f5421bcd

File: app/models/lighthouse_supporting_evidence_attachment.rb
Comment (NB28VT) [reply app/models/lighthouse_supporting_evidence_attachment.rb:5]: It is, yes

File: spec/controllers/v0/upload_supporting_evidences_controller_spec.rb
Comment (NB28VT) [reply spec/controllers/v0/upload_supporting_evidences_controller_spec.rb:1]: Yup, this was intentional, as noted in the commit message for https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/cb9e719882284ae63a68801ba0b6fc56b6163af2, I refactored this controller to remove a dependency on the `FormAttachmentCreate` module, which was a sort of poorly thought out way to DRY up some of the upload controllers shared functionality. That dependency made some of my refactors difficult, and in conversations with @SamStuckey we settled on just duping some of that code here and refactoring it to our liking.

TLDR the tests in this file were really written around that old interface and this was better tested by updating `spec/requests/upload_supporting_evidence_request_spec.rb`, the request spec you see below.

File: .github/CODEOWNERS
Comment (NB28VT) [reply .github/CODEOWNERS:1048]: Thanks dudes, looks like [I got this figured out](https://github.com/department-of-veterans-affairs/vets-api/actions/runs/7010987122/job/19072686458?pr=13835)

File: lib/lighthouse/benefits_documents/form_526_lighthouse_documents_service.rb
Comment (NB28VT) [reply lib/lighthouse/benefits_documents/form_526_lighthouse_documents_service.rb:10]: @SamStuckey quick follow up here, TLDR starting to plan out my monitoring dashboards for this release and wondering if there are any additional changes to the monitoring here that may make that easier. Still learning how team 1 has been setting up these dashboards but wondering what your experience has been/maybe we set one up together if that makes more sense thanks!

File: spec/sidekiq/evss/disability_compensation_form/upload_bdd_instructions_spec.rb
Comment (NB28VT) [reply spec/sidekiq/evss/disability_compensation_form/upload_bdd_instructions_spec.rb:72]: Good call, [went through and standardized the comments I added here](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/eb713e624767f5195752ffa9ed6e5011a2bec543), think they help a lot in these specs

**Comment by NB28VT** [comment]
> 1 Error
> 🚫	This PR changes `695` LoC (not counting whitespace/newlines).
> In order to ensure each PR receives the proper attention it deserves, those exceeding `500` will not be reviewed, nor will they be allowed to merge. Please break this PR up into smaller ones.
> 
> If you have reason to believe that this PR should be granted an exception, please see the [Submitting pull requests for approval - FAQ](https://depo-platform-documentation.scrollhelp.site/developer-docs/Submitting-pull-requests-for-approval.655032351.html#Submittingpullrequestsforapproval-FAQ).
> 
> File Summary
> Big PRs are difficult to review, often become stale, and cause delays.
> 
> Generated by 🚫 [Danger](https://danger.systems/)

Addressing this, I believe this PR is an acceptable length because:
1. Most of the changes are required for this to work end to end (for example, the duplication of existing form attachment infrastructure to escape the inheritance problem explained in [that commit](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/7e5c4658bb3877539f4f9fcad82cae95f5421bcd) is necessary to make this work)
2. Several of the changes, notably those in [the upload request spec](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/files#diff-820f02d4c9caedfefc561689af5fc11faecb4a9ad1041fbe2b055ec95b4251a9) were enforced by the linting rules even though I didn't technically add the code originally
3. It's not egregiously longer than the limit and it accomplishes the entire release for this feature migration.

An argument could be made for ditching the changes in two of the the three jobs this touches: SubmitUploads, SubmitForm0781 and UploadBDDInstructions. Those files all share my [new documents service](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/files#diff-5eb9d50f65ee2c62ff71143d96c8d73c537ed0c598e05fc4e42b78ce6d7f6dfe). We could release this with only the changes in SubmitUploads for example, but given its the same shared code that feels kind of arbitrary to follow the code length rule. If the work is done and the PR is understandable why undo all the work?

File: app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb
Comment (freeheeling) [inline app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb:8]: Should this method name be `store_dir`?

File: app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb
Comment (freeheeling) [inline app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb:34]: Are Lighthouse uploads stored to the same directory as EVSS supporting evidence attachments? Is the `guid` for each type unique?

File: app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb
Comment (NB28VT) [reply app/uploaders/lighthouse_supporting_evidence_attachment_uploader.rb]: Yeah good catch I had copied this over from the files we duped and missed this was doing basically what 3PL was doing. Fixed in https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/44b3e969af1478875b47a970db52bbc6e5e4f7a4

File: app/controllers/v0/upload_supporting_evidences_controller.rb
Comment (tommasina-va) [inline app/controllers/v0/upload_supporting_evidences_controller.rb:25]: @NB28VT would you mind explaining why this `create` isn't feature-gated? Also, is the intention to have it create both EVSS and LH supporting evidence attachment before the flag is flipped?

**Comment by mchae-nava** [comment]
Just a general comment, it'd be good to consider the ease at which we can retire the "EVSS" portion of this when the time comes. Converting is a good approach to this, so updating workers is good imo

For example, consider parts of this that may have to eventually be refactored vs removed when we reach retirement step

File: app/controllers/v0/upload_supporting_evidences_controller.rb
Comment (mchae-nava) [reply app/controllers/v0/upload_supporting_evidences_controller.rb:25]: Piggybacking on this comment

> Creates a new LighthouseSupportingEvidenceAttachment class that inherits from FormAttachment. This was required to solve an inheritance issue that is well documented in the [commit message](https://github.com/department-of-veterans-affairs/vets-api/pull/13835/commits/7e5c4658bb3877539f4f9fcad82cae95f5421bcd) for this change. Note I also had to create Lighthouse verisons of the superclasses and associated services for this attachment subtype, which mirrors the behavior in SupportingEvidenceAttachment

Can you elaborate on why a new model was needed? I see you link to the commit message, but I can't view it

File: spec/requests/upload_supporting_evidence_request_spec.rb
Comment (tommasina-va) [inline spec/requests/upload_supporting_evidence_request_spec.rb:33]: Generally, `v0` has been associated with EVSS. Maybe consider creating a new controller. (Just as an example, here's what we did for Intent To File: [EVSS](https://github.com/department-of-veterans-affairs/vets-api/blob/fe39a1c2f98ca4b685d2c47e6cae6ebeba88d3ca/app/controllers/v0/intent_to_files_controller.rb#L8); [LH](https://github.com/department-of-veterans-affairs/vets-api/blob/fe39a1c2f98ca4b685d2c47e6cae6ebeba88d3ca/modules/claims_api/app/controllers/claims_api/v2/veterans/intent_to_file_controller.rb#L9). This also makes it easier to track during the migration process, and also might help with your logging concern.

File: app/controllers/v0/upload_supporting_evidences_controller.rb
Comment (tommasina-va) [inline app/controllers/v0/upload_supporting_evidences_controller.rb:62]: I might be missing something, but would it be possible to plan out a data migration once this gets merged instead of double-creating attachments now? It seems like file naming convention is identical, so I'm not clear on why this is needed now.

**Review by tommasina-va — DISMISSED** [review]
Looks good--we know how much work goes into this migration effort! We had a few questions, mainly around the controller and creation of a new attachment class. 

We are also concerned about the general ease of retirement of the EVSS methods when the LH functionality is locked into place with them. 

We're impressed that you've tested within the LH sandbox, so we feel comfortable approving.  🥇 Nice work!

**Comment by NB28VT** [comment]
> Just a general comment, it'd be good to consider the ease at which we can retire the "EVSS" portion of this when the time comes. Converting is a good approach to this, so updating workers is good imo
> 
> For example, consider parts of this that may have to eventually be refactored vs removed when we reach retirement step. Feature gating at the top might be a good way around this, for example, but I also may be ignorant of the challenges you encountered

This was the reason I ended up creating [a service]( https://github.com/department-of-veterans-affairs/vets-api/pull/13835/files#diff-5eb9d50f65ee2c62ff71143d96c8d73c537ed0c598e05fc4e42b78ce6d7f6dfe) with single logged `upload_lighthouse_document` method and inserting that at the highest level closest to the feature gate in all of the jobs affected. So when we remove the feature gate, all we need to do is remove the EVSS code in each of those jobs that is called after the gate and the Lighthouse behavior should be neatly slotted in place.

**Comment by NB28VT** [comment]
@tommasina-va @mchae-nava I did end up removing the need for a second type of `FormAttachment`, and cleaned up the workers a bit so it will be a bit easier to sunset the EVSS code at some point (I think)

Still an open question whether we want to keep the other changes I made to the controller (removing the dependence on the `FormAttachmentCreate` concern). But I think this addresses your main questions.

Note I'm putting development here on hold as we've been asked to implement a polling solution for tracking the Lighthouse uploads after the fact. Sam is working on some infrastructure for that that will apply to a few different use cases, so I'm switching to some other work in the meantime.

Apologies if the above comments are a little hard to follow, wrote this at EOD, happy to sync up on a call next week if it's unclear where this branch ended up and if it satisfies your concerns.

**Comment by NB28VT** [comment]
Note I also need to follow up on the CI failures, the PII stuff should have been taken care of as I worked with Eric Boehs to get this branch in a safe state with those changes today, not sure why that is failing the CI check but will investigate

**Pull request closed without merging**
